Added documentation for the documentation generator (#83)
NUbots/NUbook: src/book/02-system/01-foundations/02-build-system.mdx

| ----------------- | --------------------------------------------------------------------------- |
| build             | Build the currently configured code                                         |
| configure         | Configure the CMake system for the current target                           |
| edit              | Edit configuration files                                                    |
| footdown          | Trains a neural network to detect foot down based on leg loads              |
| format            | Applies code formatting rules to the codebase                               |
| install           | Install built code to the specified robot                                   |
| module            | Generate new modules                                                        |
| nbs               | For working with NBS recordings                                             |
| run               | Run a role locally                                                          |
| shell             | Exposes a shell in the Docker image                                         |
| target            | Select a target to work on                                                  |

> | extract-reactions | Dump documentation of each module's ons, emits and reactions in JSON format |

Reviewer: How so?

| run       | Run a role locally                                             |
| shell     | Exposes a shell in the Docker image                            |
| target    | Select a target to work on                                     |

> | doc       | Create json documenting each module's reactors                 |

Reviewer 2: ```suggestion
| doc       | Dump documentation of each module's emits and reactions in JSON format                 |
```

Reviewer 2: The rest of the table might need formatting if you commit this suggestion.

The `target` command will download (or, if needed, build) the Docker image for the specified target and then mark that target as active.

> ##### Doc

Reviewer 2: Just spit balling, but would something like `./b extract-reactions` or similar be a good alternative. That way we can keep the `./b doc` to be something else, like extract READMEs and comments from modules?

> The `extract-reactions` command of `b` generates a JSON file documenting each Module. It lists each reactor, with their on statements and emit statements. The on and emit statements have their location in the file and their dsl worlds also.

Reviewer: How so?

> The `extract-documentation` command of `b` generates a JSON file documenting each Module. It lists each reactor, with their on statements and emit statements. The on and emit statements have their location in the file and their dsl worlds also.

Reviewer 2: ```suggestion
The `./b extract-documentation` command generates a JSON file documenting each module. It lists each reactor, with their on statements and emit statements. The on and emit statements have their location in the file and their DSL words also.
```

The `--indir` argument gives the root of the directories that will be documented, defaults to the `./module/` directory.

The `--outdir` argument gives the directory that the JSON output will be written, defaults to `./doc/NUdoc/` directory.

> The `-m` or `--multiprocess` argument tells the command to use multiprocessing to greatly reduce the processing time, but uses a very large amount of RAM, about 1G per thread.

Reviewer 2: ```suggestion
The `-m` or `--multiprocess` argument tells the command to use multiprocessing to greatly reduce the processing time, but uses a very large amount of RAM, about 1GB per thread.
```

> The `-m` or `--multiprocess` argument tells the command to use multiprocessing to greatly reduce the processing time, but uses a very large amount of RAM, about 1GB per thread.

Reviewer: Readded

> 

Reviewer 2: The plan is to eventually extract the documentation, generate the images, and submit a PR to NUbook. Ideally automatically in CI when NUbots code changes. 

The JSON here is custom, and I think Cameron has written something that does the conversion.

> 

Reviewer: Argparse auto generates a `-h` argument, I don't think it needs to be explicitly documented. If it should exist it should be general for all b scripts.

> 

Reviewer: @KipHamiltons Have you found any clang api docs? I haven't.
If you need to convert to DOT https://gist.github.com/ElderNoSpace/f0042e99b713601e0589a7692f6bff75
I don't think this tool warrants its own page.

> 

Reviewer 3: Here is a couple of links which would perhaps be useful? I'll let you be the judge of whether they give good insight into using the python bindings.
https://libclang.readthedocs.io/en/latest/
https://clang.llvm.org/doxygen/group__CINDEX.html
https://eli.thegreenplace.net/2011/07/03/parsing-c-in-python-with-clang
